### PR TITLE
ci: semantic-release

### PR DESCRIPTION
**Pull request by gr2m**
Follow up to https://github.com/node-fetch/node-fetch/pull/1257#issuecomment-909367405

This pull request adds [semantic-release](https://semantic-release.gitbook.io/semantic-release/). Once merged, GitHub/npm releases will be automated based on the following commit message conventions:

1. `fix: ...` or `fix(scope name): ...` prefix in subject: bumps fix version, e.g. `1.2.3` → `1.2.4`
2. `feat: ...` or `feat(scope name): ...` prefix in subject: bumps feature version, e.g. `1.2.3` → `1.3.0`
3. `BREAKING CHANGE:` in body: bumps breaking version, e.g. `1.2.3` → `2.0.0`

Only one version number is bumped at a time, the highest version change trumps the others.

Besides publishing a new version to npm, semantic-release also creates a git tag and release on GitHub, generates changelogs from the commit messages and puts them into the release notes, creates comments on pull requests included in the release as well as the issues they resolved to notify about the new release.

`semantic-release` with the current configruation supports releases to the `@next` channel and pre-relaeses using the `next` and `beta` branches. There are some gotchas here, but I'd say we get there when we get there.

I've been the maintainer of `semantic-release` for many years now, and I use it on nearly [all my npm modules](https://www.npmjs.com/~gr2m), I couldn't imagine doing Open Source without it. Let me know if you have any questions.

### Todos

- [x] #1271
- [ ] #1272
- [x] Add `NPM_TOKEN` secret to the repository. I recommend creating a dedicated npm account that can only publish the `node-fetch` package. The bot account can have 2FA enabled, npm has automation tokens now that will work with semantic-release.
- [x] Document the commit conventions above somewhere. I'd usually add them to `CONTRIBUTING.md`, but we don't have one yet :)
- [x] Setup branch protection for `v2.x`, `beta`, `next` branches.

I would recommend

- [x] Disable all merge strategies but `squash & merge`. Do not ask contributors to fix their commit messages, we can just squash them into a single commit when merging.
- [x] Install the [`semantic-pull-request` app](https://github.com/zeke/semantic-pull-requests) 
- [x] Configure `semantic-pull-request` with [`titleOnly: true`](https://github.com/zeke/semantic-pull-requests#configuration).

### CODE REVIEW

**Comment by jimmywarting** [comment]
It sounds awesome, think i would like to incorporate it into my repos also

**Comment by jimmywarting** [comment]
Ping @bitinn

**Review by jimmywarting — APPROVED** [review]

**Comment by jimmywarting** [comment]
ping @Richienb @xxczaki to take a look at this.

(Also could i have admin privileges?)

**Comment by xxczaki** [comment]
> ping @Richienb @xxczaki to take a look at this.
> 
> (Also could i have admin privileges?)

Sorry, the last couple of months have been quite rough for me in terms of work. I've given you the admin privileges. When it comes to this PR, I will take a look at it this week.

**Comment by jimmywarting** [comment]
I created a dedicated npm account that can only publish the node-fetch package. it has 2FA enabled.

I have also generated a NPM_TOKEN and added it to the repository secret

**Comment by jimmywarting** [comment]
@gr2m what is left to do?

**Comment by gr2m** [comment]
I'll add a `CONTRIBUTING.md` and document the commit conventions there for maintainers. Then we need to do branch protections, I no longer have admin rights to this repository so I cannot do that myself. And I'd disable the "Rebase and merged" option to merge pull requests.

However, when merging `next` into `main`, we have to use "Create a merge commit". I usually enable the option specifically when needed only for this use case. I'll document it in `CONTRIBUTING.md`

**Comment by gr2m** [comment]
I'm set on my side, I do not have permission to do the rest. Feel free to make me an admin and I'll take care of the rest. I think I used to have the access rights, but they might have gotten lost with the transition to @node-fetch? I'm okay either way, let me know if you have any questions.

Once we are done here, I'll backport the setup to `2.x`

**Comment by jimmywarting** [comment]
I gave u admin rights

**Comment by gr2m** [comment]
@jimmywarting can you do a final review? I added `CONTRIBUTING.md` with instructions on how to release new versions

**Review by jimmywarting — APPROVED** [review]
LGTM

File: package.json [hunk 1]
```diff
@@ -116,5 +116,16 @@
       }
     ]
   },
-  "runkitExampleFilename": "example.js"
+  "runkitExampleFilename": "example.js",
+  "release": {
+    "branches": [
+      "+([0-9]).x",
+      "main",
+      "next",
```
Comment (LinusU) [inline package.json:124]: Does `next` need any special config, like `beta`, or it is automatically picked up from the name?

**Review by LinusU — APPROVED** [review]
LGTM 👍

File: package.json (hunk 1 above)
Comment (gr2m) [reply package.json:124]: it's automatically picked up from the name

**Comment by jimmywarting** [comment]
🎉  Merged 🎉 

hope this makes more rapid releases now
guess we can go ahead and delete that changelog.md now

**Comment by gr2m** [comment]
w00p :)

Don't hesitate to ping me for any questions, and in case something is acting up 

> guess we can go ahead and delete that changelog.md now

Yes, let's do that. I usually link to the latest version of the `changelog.md` from the first release notes that are created by semantic-release

**Pull request merged**